yateto/codegen/cache.py

```diff
@@ -18,11 +22,18 @@ def addRoutine(self, name, generator):
     if generatorName not in self._generators:
       self._generators[generatorName] = generator
   
-  def generate(self, header, cppFileName):
-    with Cpp(cppFileName) as cpp:
-      for generator in self._generators.values():
-        generator.header(cpp)
+  def generate(self, header, cppFileName, gpuFileName):
+    with Cpp(gpuFileName) as gpucpp:
+      with Cpp(cppFileName) as cpp:
+        for generator in self._generators.values():
+          if isinstance(generator, GpuRoutineGenerator):
+            generator.header(gpucpp)
+          else:
+            generator.header(cpp)
 
     for name, generator in self._routines.items():
-      declaration = generator(name, cppFileName)
-      header(declaration)
+      if isinstance(generator, GpuRoutineGenerator):
+        declaration = generator(name, gpuFileName)
+      else:
+        declaration = generator(name, cppFileName)
+      header(declaration)
```

Comment:
Not a fan of this pattern - this is bad OOP design. No need to change now, but please keep this in mind in the future.